python/dgl/distributed/optim/pytorch/sparse_optim.py

```diff
@@ -0,0 +1,221 @@
+        '''
+        with th.no_grad():
+            local_indics = {emb.name: [] for emb in self._params}
+            local_grads = {emb.name: [] for emb in self._params}
+            device = th.device('cpu')
+            for emb in self._params:
+                name = emb._tensor.name
+                kvstore = emb._tensor.kvstore
+                trace = emb._trace
+                trainers_per_server = self._world_size // kvstore.num_servers
+
+                idics = [t[0] for t in trace]
+                grads = [t[1].grad.data for t in trace]
+                idics = th.cat(idics, dim=0)
+                grads = th.cat(grads, dim=0)
+                device = grads.device
+
+                # will send grad to each corresponding trainer
+                if self._world_size > 1:
+                    # get idx split from kvstore
+                    idx_split = kvstore.get_partid(name, idics)
+                    idx_split_size = []
+                    idics_list = []
+                    grad_list = []
+                    # split idx and grad first
+                    for i in range(kvstore.num_servers):
+                        mask = idx_split == i
+                        idx_i = idics[mask]
+                        grad_i = grads[mask]
+
+                        kv_idx_split = th.remainder(idx_i, trainers_per_server).long()
+                        for j in range(trainers_per_server):
+                            mask = kv_idx_split == j
+                            idx_j = idx_i[mask]
+                            grad_j = grad_i[mask]
+                            idx_split_size.append(th.tensor([idx_j.shape[0]], dtype=th.int64))
+                            idics_list.append(idx_j)
+                            grad_list.append(grad_j)
+
+                    # if one machine launch multiple KVServer, they share the same storage.
+                    # For each machine, the pytorch rank is num_trainers * machine_id + i
+
+                    # use scatter to sync across trainers about the p2p tensor size
+                    # Note: If we have GPU nccl support, we can use all_to_all to sync information here
+                    gather_list = list(th.empty([self._world_size], dtype=th.int64).chunk(self._world_size))
+                    alltoall_cpu(self._rank, self._world_size, gather_list, idx_split_size)
+                    # use cpu until we have GPU alltoallv
+                    idx_gather_list = [th.empty((int(num_emb),), dtype=idics.dtype) for num_emb in gather_list]
+                    alltoallv_cpu(self._rank, self._world_size, idx_gather_list, idics_list, idics.dtype)
+                    local_indics[name] = idx_gather_list
+                    grad_gather_list = [th.empty((int(num_emb), grads.shape[1]), dtype=grads.dtype) for num_emb in gather_list]
+                    alltoallv_cpu(self._rank, self._world_size, grad_gather_list, grad_list, idics.dtype)
+                    local_grads[name] = grad_gather_list
```

Comment:
ideally, the entire thing should be done in C. but let's see how much overhead it has first.